File: 卡码/代码随想录/1/09-动态规划/Js/05-子序列问题/04-回文/LeetCode_0516.js
//516. 最长回文子序列
const longestPalindromeSubseq = function(s) {
	const n=s.length;
	const dp=new Array(n);
	for(let i=0;i<n;i++){
		dp[i]=new Array(n).fill(0);
		dp[i][i]=1;
	}
	for(let i=n-1;i>=0;i--){
		for(let j=i+1;j<n;j++){
			if(s[i]==s[j]) dp[i][j]=dp[i+1][j-1]+2;
			else dp[i][j]=Math.max(dp[i+1][j],dp[i][j-1])
		}
	}
	return dp[0][n-1];
};

File: 卡码/代码随想录/1/09-动态规划/Go/05-子序列问题/04-回文/LeetCode_0516.go
// 516. 最长回文子序列
package main

import (
	"math"
)

func longestPalindromeSubseq(s string) int {
	n := len(s)
	dp := make([][]int, n)
	for i := 0; i < n; i++ {
		dp[i] = make([]int, n)
		dp[i][i] = 1
	}
	for i := n - 1; i >= 0; i-- {
		for j := i + 1; j < n; j++ {
			if s[i] == s[j] {
				dp[i][j] = dp[i+1][j-1] + 2
			} else {
				dp[i][j] = int(math.Max(float64(dp[i][j-1]), float64(dp[i+1][j])))
			}
		}
	}
	return dp[0][n-1]
}
